target/snitch_cluster/sw/apps/snax-mac-simple/src/snax-mac-simple.c

```diff
@@ -0,0 +1,81 @@
+// Copyright 2020 ETH Zurich and University of Bologna.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "snrt.h"
+
+#include "data.h"
+
+int main() {
+    // Set err value for checking
+    int err = 0;
+
+    uint64_t *local_a, *local_b;
+    uint64_t *local_o;
+
+    // Allocate space in TCDM
+    local_a = (uint64_t *)snrt_l1_next();
+    local_b = local_a + VEC_LEN;
+    local_o = local_b + VEC_LEN;
+
+    uint32_t dma_pre_load = snrt_mcycle();
+
+    // Use data mover core to bring data from L3 to TCDM
+    if (snrt_is_dm_core()) {
+        size_t vector_size = VEC_LEN * sizeof(uint64_t);
+        snrt_dma_start_1d(local_a, A, vector_size);
+        snrt_dma_start_1d(local_b, B, vector_size);
+    }
+
+    // Wait until DMA transfer is done
+    snrt_cluster_hw_barrier();
+
+    // Read the mcycle CSR (this is our way to mark/delimit a specific
+    // code region for benchmarking)
+    uint32_t pre_is_compute_core = snrt_mcycle();
+
+    if (snrt_is_compute_core()) {
+        // This marks the start of the accelerator style of MAC operation
+        uint32_t csr_set = snrt_mcycle();
+
+        // Set addresses
+        write_csr(0x3d0, (uint64_t)local_a);
+        write_csr(0x3d1, (uint64_t)local_b);
+        write_csr(0x3d3, (uint64_t)local_o);
+
+        // Set configs
+        write_csr(0x3d4, 1);   // Number of iterations
+        write_csr(0x3d5, 19);  // Vector length
+        write_csr(0x3d6, 1);   // Set simple multiplication
+
+        // Write start CSR to launch accelerator
+        write_csr(0x3c0, 0);
```

Comment:
Maybe we can also use a header here, then we can make the code way more readable!:
```c
        write_csr(MAC_ADDRESS_A, (uint64_t)local_a);
        write_csr(MAC_ADDRESS_B, (uint64_t)local_b);
        write_csr(MAC_ADDRESS_D, (uint64_t)local_o);

        // Set configs
        write_csr(MAC_NB_ITER, 1);   // Number of iterations
        write_csr(MAC_VEC_LEN, 19);  // Vector length
        write_csr(MAC_SIMPLE_MODE, 1);   // Set simple multiplication

        // Write start CSR to launch accelerator
        write_csr(SNAX_LAUNCH, 0);
```
The header could also be used by multiple files in the repo!
`mac_registers.h`:
```c
#define MAC_ADDRESS_A 0x3d0
```

Reply:
Oh okay! I will add this thanks :) so that it's clearer too!!

Reply:
Hmm so the compiler does not like this. It appears because the `read_csr` and `write_csr` were also `#define` functions...

See the [`write_csr`](https://github.com/pulp-platform/riscv-opcodes/blob/dbb086ae703b2fccbaa79036e11802e206b27e46/encoding.h#L248). It looks like the toolchain requires that we explicitly put the code into the compiler. Otherwise it complains things like:

```bash
rc/snax-mac.c:48:9: error: operand must be a valid system register name or an integer in the range [0, 4095]
        write_csr(MAC_ADDRESS_A, (uint64_t)local_a);
        ^
/repo/sw/snRuntime/api/../../deps/riscv-opcodes/encoding.h:253:17: note: expanded from macro 'write_csr'
  asm volatile ("csrw " #reg ", %0" :: "rK"(val)); })
                ^
<inline asm>:1:7: note: instantiated into assembly here
        csrw MAC_ADDRESS_A, a0
             ^
1 error generated.
```

Reply:
Bump! @JosseVanDelm

Reply:
seems like it should work?
https://stackoverflow.com/questions/60141112/get-define-value-in-preprocessor-time-in-c
Can you make sure non-macros can access those defined values?

Reply:
Chatgpt has spoken!

> In C, you can use macro constants within macro functions by directly referencing the macro constants within the macro definition. When you define a macro function using #define, **the macro constants are treated as literal text and are substituted directly into the expanded code**.